gobblin-runtime/src/main/java/gobblin/runtime/MultiConverter.java

```diff
@@ -80,87 +79,122 @@ public Object convertSchema(Object inputSchema, WorkUnitState workUnit)
   }
 
   /**
-   * A type of {@link java.util.Iterator} to be used with {@link MultiConverter}.
+   * A type of {@link java.util.Iterator} to be used with {@link MultiConverter}. The Converter uses the
+   * {@link ChainedConverterIterator} to chain iterators together. The first ChainedConverterIterator created contains
+   * an iterator with only the inputRecord and the first converter in the converters list. Each subsequent
+   * ChainedConverterIterator is created using the previous ChainedConverterIterator along with the next converter in
+   * the converters list. By chaining the converters and iterators in this fashion, a reference to the last
+   * ChainedConverterIterator will be sufficient to iterate through all the data.
    */
   private class MultiConverterIterator implements Iterator<Object> {
 
     private final WorkUnitState workUnitState;
-    private final Deque<Iterator<Object>> converterIteratorStack = Lists.newLinkedList();
-    private Iterator<Object> outputIterator;
+
+    private Iterator<Object> chainedConverterIterator;
 
     public MultiConverterIterator(Object inputRecord, WorkUnitState workUnitState) throws DataConversionException {
       this.workUnitState = workUnitState;
-
-      // Construct the initial stack of converter iterators
-      if (converters.isEmpty()) {
-        this.converterIteratorStack.push(new SingleRecordIterable<Object>(inputRecord).iterator());
-      } else {
-        Object record = inputRecord;
-        for (Converter converter : converters) {
-          if (!this.converterIteratorStack.isEmpty()) {
-            record = this.converterIteratorStack.peek().next();
-          }
-          Iterator<Object> iterator =
-              converter.convertRecord(convertedSchemaMap.get(converter), record, workUnitState).iterator();
-          this.converterIteratorStack.push(iterator);
-          // Do not continue when encountering an empty iterator
-          if (!iterator.hasNext()) {
-            break;
-          }
+      this.chainedConverterIterator = new SingleRecordIterable<Object>(inputRecord).iterator();
+
+      for (int i = 0; i < converters.size(); i++) {
+        if (i == 0) {
```

Comment:
Originally it was like that so that in case the "converters" list was empty it would still work. I still implemented your suggestions by creating an IdentityConverter class